File: hackathon/src/main/java/com/pi/hackathon/service/HackathonService.java
package com.pi.hackathon.service;

import com.pi.hackathon.entities.Hackathon;
import com.pi.hackathon.entities.HackathonParticipation;
import com.pi.hackathon.repository.HackathonRepository;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.Set;

@Service
@AllArgsConstructor
public class HackathonService {
    private final HackathonRepository hackathonRepository;

    public Hackathon createHackathon(Hackathon hackathon) {
        return hackathonRepository.save(hackathon);
    }

    // Get a Hackathon by ID
    public Optional<Hackathon> getHackathonById(int id) {
        return hackathonRepository.findById(id);
    }

    // Get all Hackathons with pagination
    public Page<Hackathon> getAllHackathons(int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return hackathonRepository.findAll(pageable);
    }

    // Update an existing Hackathon
    public Hackathon updateHackathon(int id, Hackathon updatedHackathon) {
        return hackathonRepository.findById(id).map(hackathon -> {
            hackathon.setTitle(updatedHackathon.getTitle());
            hackathon.setDescription(updatedHackathon.getDescription());
            hackathon.setLocation(updatedHackathon.getLocation());
            hackathon.setStartTime(updatedHackathon.getStartTime());
            hackathon.setEndTime(updatedHackathon.getEndTime());
            hackathon.setNbrPlaces(updatedHackathon.getNbrPlaces());
            hackathon.setNiveauDifficulte(updatedHackathon.getNiveauDifficulte());
            hackathon.setNiveauImportance(updatedHackathon.getNiveauImportance());
            return hackathonRepository.save(hackathon);
        }).orElseThrow(() -> new RuntimeException("Hackathon not found"));
    }

    // Delete a Hackathon
    public void deleteHackathon(int id) {
        hackathonRepository.deleteById(id);
    }

    // Get all participants of a Hackathon
    public List<HackathonParticipation> getParticipants(int hackathonId) {
        return hackathonRepository.findById(hackathonId)
                .map(Hackathon::getParticipations)
                .orElseThrow(() -> new RuntimeException("Hackathon not found"));
    }
}

File: src/app/services/hackathon.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Hackathon } from '../models/hackathon';
import { HackathonParticipation } from '../models/hackathon-participation';

@Injectable({
  providedIn: 'root',
})
export class HackathonService {
  private baseUrl = 'http://localhost:8222/api/v1/hackathons';

  constructor(private http: HttpClient) {}

  // Create Hackathon
  createHackathon(hackathon: Hackathon): Observable<Hackathon> {
    return this.http.post<Hackathon>(`${this.baseUrl}`, hackathon);
  }

  // Get Hackathon by ID
  getHackathonById(id: number): Observable<Hackathon> {
    return this.http.get<Hackathon>(`${this.baseUrl}/${id}`);
  }

  // Get All Hackathons with Pagination
  getAllHackathons(page: number = 0, size: number = 10): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}?page=${page}&size=${size}`);
  }

  // Update Hackathon
  updateHackathon(id: number, hackathon: Hackathon): Observable<Hackathon> {
    return this.http.put<Hackathon>(`${this.baseUrl}/${id}`, hackathon);
  }

  // Delete Hackathon
  deleteHackathon(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // Get Hackathon Participants
  getHackathonParticipants(id: number): Observable<HackathonParticipation[]> {
    return this.http.get<HackathonParticipation[]>(
      `${this.baseUrl}/${id}/participants`
    );
  }
}
